presto-spi/src/main/java/com/facebook/presto/spi/SubfieldPath.java

```diff
@@ -75,59 +117,118 @@ public boolean equals(Object o)
             if (o == null || getClass() != o.getClass()) {
                 return false;
             }
-            PathElement other = (PathElement) o;
-            if (isSubscript != other.isSubscript) {
-                return false;
-            }
-            if (field == null && other.field == null) {
-                return subscript == other.subscript;
+
+            LongSubscript that = (LongSubscript) o;
+            return index == that.index;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(index);
+        }
+
+        @Override
+        public String toString()
+        {
+            return "[" + index + "]";
+        }
+
+        @Override
+        public boolean isSubscript()
+        {
+            return true;
+        }
+    }
+
+    public static final class StringSubscript
+            extends PathElement
+    {
+        private final String index;
+
+        @JsonCreator
+        public StringSubscript(@JsonProperty("index") String index)
+        {
+            this.index = requireNonNull(index, "index is null");
+        }
+
+        @JsonProperty
+        public String getIndex()
+        {
+            return index;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
             }
-            if (field == null || other.field == null) {
+            if (o == null || getClass() != o.getClass()) {
                 return false;
             }
-            return field.equals(other.field);
+
+            StringSubscript that = (StringSubscript) o;
+            return Objects.equals(index, that.index);
         }
 
         @Override
         public int hashCode()
         {
-            return field != null ? field.hashCode() : (int) subscript;
+            return Objects.hash(index);
         }
 
         @Override
         public String toString()
         {
-            if (field != null) {
-                return field;
-            }
-            return Long.valueOf(subscript).toString();
+            return "[" + index + "]";
+        }
+
+        @Override
+        public boolean isSubscript()
+        {
+            return true;
         }
     }
 
-    private final ArrayList<PathElement> path;
+    private final String name;
+    private final List<PathElement> path;
 
+    public static PathElement allSubscripts()
```

Comment:
@tdcmeehan Let's say you have column coordinates of type array(row(x, y)) and the query is:

```
SELECT c.x FROM t CROSS JOIN UNNEST(coordinates) as u(c)
```

By default, table scan will produce coordinates column where each array element contains both x and y. However, the query only uses x. To express that, you'll use SubscriptPath `coordinates[*].x`. allSubscripts is used to express `*`.

I can't come up with a good comment to put here. Any ideas? I'd like to avoid restating the obvious, e.g. "Represents all subscripts of an array or map."

Reply:
What do you think of expressing all subscripts as its own subclass of `PathElement`, rather than overloading `LongSubscript` with the magic number?

Reply:
@tdcmeehan I think is it better. We can then serialize it as "*" which will read better. Would you like me to make this change in this PR?

Reply:
Sure! 👍